File: src/adapters/iflow/install_iflow_integration.py
"""
iFlow CLI Hook集成安装脚本
为iFlow CLI安装跨CLI协作感知能力

使用方法：
python install_iflow_integration.py [--verify|--uninstall]
"""

import os
import sys
import json
import shutil
import argparse
from pathlib import Path
from datetime import datetime

# 获取当前文件目录
current_dir = Path(__file__).parent
project_root = current_dir.parent.parent.parent

# iFlow CLI配置路径
IFLOW_CONFIG_DIR = os.path.expanduser("~/.config/iflow")
IFLOW_HOOKS_FILE = os.path.join(IFLOW_CONFIG_DIR, "hooks.yml")

def create_iflow_config_directory():
    """创建iFlow配置目录"""
    os.makedirs(IFLOW_CONFIG_DIR, exist_ok=True)
    print(f"✅ 创建iFlow配置目录: {IFLOW_CONFIG_DIR}")

def install_iflow_hooks():
    """安装iFlow Hook配置"""
    # 读取现有hooks配置
    existing_hooks = {}
    if os.path.exists(IFLOW_HOOKS_FILE):
        try:
            import yaml
            with open(IFLOW_HOOKS_FILE, 'r', encoding='utf-8') as f:
                existing_hooks = yaml.safe_load(f) or {}
        except Exception as e:
            print(f"⚠️ 读取现有hooks配置失败: {e}")
            existing_hooks = {}

    # 定义跨CLI协作的Hook配置
    cross_cli_hooks = {
        "cross_cli_hook_adapter": {
            "name": "CrossCLIHookAdapter",
            "module": "src.adapters.iflow.hook_adapter",
            "class": "IFlowHookAdapter",
            "enabled": True,
            "priority": 100,
            "hooks": [
                "on_command_start",
                "on_command_end",
                "on_user_input",
                "on_workflow_stage",
                "on_pipeline_execute",
                "on_output_render",
                "on_error"
            ],
            "config": {
                "cross_cli_enabled": True,
                "supported_clis": ["claude", "gemini", "qwencode", "qoder", "codebuddy", "copilot"],
                "auto_detect": True,
                "timeout": 30,
                "error_handling": "continue",
                "collaboration_mode": "active"
            }
        }
    }

    # 合并配置（保留现有hooks，添加协作功能）
    merged_hooks = existing_hooks.copy()
    if 'plugins' not in merged_hooks:
        merged_hooks['plugins'] = []

    # 检查是否已存在跨CLI Hook
    existing_plugin_names = [plugin.get('name') for plugin in merged_hooks.get('plugins', [])]
    cross_cli_hook_exists = any(plugin_name == 'CrossCLIHookAdapter' for plugin_name in existing_plugin_names)

    if not cross_cli_hook_exists:
        merged_hooks['plugins'].append(cross_cli_hooks['cross_cli_hook_adapter'])

    # 写入hooks配置文件
    try:
        import yaml
        with open(IFLOW_HOOKS_FILE, 'w', encoding='utf-8') as f:
            yaml.dump(merged_hooks, f, default_flow_style=False, allow_unicode=True)

        print(f"✅ iFlow Hook配置已安装: {IFLOW_HOOKS_FILE}")
        print("🔗 已安装的Hook:")
        for plugin in merged_hooks.get('plugins', []):
            if plugin.get('name') == 'CrossCLIHookAdapter':
                print(f"   - {plugin['name']}: ✅ 跨CLI协作感知")
                print(f"     支持的CLI: {', '.join(plugin['config'].get('supported_clis', []))}")

        return True
    except Exception as e:
        print(f"❌ 安装iFlow Hook配置失败: {e}")
        return False

def copy_adapter_file():
    """复制适配器文件到iFlow配置目录"""
    try:
        # 创建适配器目录
        adapter_dir = os.path.join(IFLOW_CONFIG_DIR, "adapters")
        os.makedirs(adapter_dir, exist_ok=True)

        # 复制适配器文件
        adapter_files = [
            "hook_adapter.py",
            "standalone_iflow_adapter.py"
        ]

        for file_name in adapter_files:
            src_file = current_dir / file_name
            dst_file = os.path.join(adapter_dir, file_name)

            if src_file.exists():
                shutil.copy2(src_file, dst_file)
                print(f"✅ 复制适配器文件: {file_name}")
            else:
                print(f"⚠️ 适配器文件不存在: {file_name}")

        return True
    except Exception as e:
        print(f"❌ 复制适配器文件失败: {e}")
        return False

def verify_installation():
    """验证安装是否成功"""
    print("\n🔍 验证iFlow CLI集成安装...")

    # 检查配置目录
    if not os.path.exists(IFLOW_CONFIG_DIR):
        print(f"❌ 配置目录不存在: {IFLOW_CONFIG_DIR}")
        return False

    # 检查hooks文件
    if not os.path.exists(IFLOW_HOOKS_FILE):
        print(f"❌ Hooks配置文件不存在: {IFLOW_HOOKS_FILE}")
        return False

    # 检查适配器目录
    adapter_dir = os.path.join(IFLOW_CONFIG_DIR, "adapters")
    if not os.path.exists(adapter_dir):
        print(f"❌ 适配器目录不存在: {adapter_dir}")
        return False

    # 读取并验证hooks配置
    try:
        import yaml
        with open(IFLOW_HOOKS_FILE, 'r', encoding='utf-8') as f:
            hooks_config = yaml.safe_load(f) or {}

        plugins = hooks_config.get('plugins', [])
        cross_cli_plugin = None

        for plugin in plugins:
            if plugin.get('name') == 'CrossCLIHookAdapter':
                cross_cli_plugin = plugin
                print(f"✅ 跨CLI协作Hook: 已启用")
                print(f"✅ 支持的CLI工具: {', '.join(plugin.get('config', {}).get('supported_clis', []))}")
                break

        if not cross_cli_plugin:
            print("❌ 跨CLI协作Hook: 未找到")
            return False

    except Exception as e:
        print(f"❌ 读取hooks配置失败: {e}")
        return False

def uninstall_iflow_integration():
    """卸载iFlow集成"""
    try:
        # 备份现有配置
        if os.path.exists(IFLOW_HOOKS_FILE):
            backup_file = f"{IFLOW_HOOKS_FILE}.backup_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
            shutil.copy2(IFLOW_HOOKS_FILE, backup_file)
            print(f"📦 已备份现有配置: {backup_file}")

        # 移除跨CLI Hook
        if os.path.exists(IFLOW_HOOKS_FILE):
            import yaml
            with open(IFLOW_HOOKS_FILE, 'r', encoding='utf-8') as f:
                hooks_config = yaml.safe_load(f) or {}

            if 'plugins' in hooks_config:
                plugins = hooks_config['plugins']
                # 移除跨CLI Hook
                plugins = [p for p in plugins if p.get('name') != 'CrossCLIHookAdapter']

                hooks_config['plugins'] = plugins

                # 写回配置
                with open(IFLOW_HOOKS_FILE, 'w', encoding='utf-8') as f:
                    yaml.dump(hooks_config, f, default_flow_style=False, allow_unicode=True)

        print("✅ iFlow跨CLI协作集成已卸载")
        return True
    except Exception as e:
        print(f"❌ 卸载失败: {e}")
        return False

def main():
    parser = argparse.ArgumentParser(
        description="iFlow CLI跨CLI协作集成安装脚本",
        formatter_class=argparse.RawDescriptionHelpFormatter
    )

    parser.add_argument(
        "--install",
        action="store_true",
        help="安装iFlow CLI跨CLI协作集成"
    )

    parser.add_argument(
        "--verify",
        action="store_true",
        help="验证iFlow CLI集成安装"
    )

    parser.add_argument(
        "--uninstall",
        action="store_true",
        help="卸载iFlow CLI跨CLI协作集成"
    )

    args = parser.parse_args()

    print("🔄 iFlow CLI跨CLI协作集成安装器")
    print("=" * 60)

    if args.uninstall:
        print("🗑️ 卸载模式...")
        success = uninstall_iflow_integration()
    elif args.verify:
        print("🔍 验证模式...")
        success = verify_installation()
    elif args.install or len(sys.argv) == 1:
        print("📦 安装模式...")

        # 1. 创建配置目录
        print("1️⃣ 创建配置目录...")
        create_iflow_config_directory()

        # 2. 安装hooks配置
        print("2️⃣ 安装hooks配置...")
        hooks_success = install_iflow_hooks()

        # 3. 复制适配器文件
        print("3️⃣ 复制适配器文件...")
        adapter_success = copy_adapter_file()

        success = hooks_success and adapter_success

        if success:
            print("\n🎉 iFlow CLI集成安装成功！")
            print("\n📋 安装摘要:")
            print(f"   ✅ 配置目录: {IFLOW_CONFIG_DIR}")
            print(f"   ✅ Hooks文件: {IFLOW_HOOKS_FILE}")
            print(f"   ✅ 适配器目录: {os.path.join(IFLOW_CONFIG_DIR, 'adapters')}")
            print(f"   ✅ 跨CLI协作: 已启用")
            print(f"   ✅ 支持的CLI: claude, gemini, qwencode, qoder, codebuddy, copilot")
        else:
            print("\n❌ iFlow CLI集成安装失败！")
    else:
        parser.print_help()

if __name__ == "__main__":
    main()

File: src/adapters/iflow/install_iflow_integration.js
#!/usr/bin/env node

/**
 * iFlow CLI Hook Adapter Installation Script
 * Installs cross-CLI collaboration awareness capability for iFlow CLI
 * 
 * Usage:
 * node install_iflow_integration.js [--verify|--uninstall]
 */

const fs = require('fs').promises;
const path = require('path');
const os = require('os');
const yaml = require('js-yaml');

// iFlow CLI config path
const IFLOW_CONFIG_DIR = path.join(os.homedir(), '.config', 'iflow');
const IFLOW_HOOKS_FILE = path.join(IFLOW_CONFIG_DIR, 'hooks.yml');

class IFlowIntegrationInstaller {
  constructor() {
    this.startTime = Date.now();
  }

  /**
   * Create iFlow config directory
   */
  async createIFlowConfigDirectory() {
    try {
      await fs.mkdir(IFLOW_CONFIG_DIR, { recursive: true });
      console.log(`[OK] Created iFlow config directory: ${IFLOW_CONFIG_DIR}`);
      return true;
    } catch (error) {
      console.log(`[ERROR] Failed to create iFlow config directory: ${error.message}`);
      return false;
    }
  }

  /**
   * Install iFlow Hook configuration
   */
  async installIFlowHooks() {
    try {
      // Read existing hooks config
      let existingHooks = {};
      try {
        const data = await fs.readFile(IFLOW_HOOKS_FILE, 'utf8');
        existingHooks = yaml.load(data) || {};
      } catch (error) {
        console.log(`[WARN] Failed to read existing hooks config: ${error.message}`);
        existingHooks = {};
      }

      // Define cross-CLI collaboration Hook configuration
      const crossCliHooks = {
        cross_cli_adapter: {
          module: 'src.adapters.iflow.hook_adapter',
          class: 'IFlowHookAdapter',
          enabled: true,
          priority: 100,
          events: ['workflow_start', 'task_execute', 'pipeline_complete', 'user_input']
        }
      };

      // Merge config (preserve existing config, add collaboration feature)
      const mergedHooks = { ...existingHooks, ...crossCliHooks };

      // Write hooks config file
      const yamlContent = yaml.dump(mergedHooks, { indent: 2, lineWidth: 120, noRefs: true });
      await fs.writeFile(IFLOW_HOOKS_FILE, yamlContent, 'utf8');

      console.log(`[OK] iFlow Hook configuration installed: ${IFLOW_HOOKS_FILE}`);
      console.log('Installed Hooks:');
      Object.keys(crossCliHooks).forEach(hookName => {
        console.log(`   - ${hookName}: [OK] Cross-CLI collaboration awareness`);
      });

      return true;
    } catch (error) {
      console.log(`[ERROR] Failed to install iFlow Hook configuration: ${error.message}`);
      return false;
    }
  }

  /**
   * Copy adapter files to iFlow config directory
   */
  async copyAdapterFile() {
    try {
      // Create adapter directory
      const adapterDir = path.join(IFLOW_CONFIG_DIR, 'adapters');
      await fs.mkdir(adapterDir, { recursive: true });

      // Get current script directory
      const currentDir = __dirname;
      const adapterSource = path.join(currentDir, 'hook_adapter.js');
      const adapterDest = path.join(adapterDir, 'hook_adapter.js');

      // Copy adapter file
      try {
        await fs.access(adapterSource);
        await fs.copyFile(adapterSource, adapterDest);
        console.log(`[OK] Copied adapter file: ${adapterDest}`);
      } catch (error) {
        console.log(`[ERROR] Adapter source file does not exist: ${adapterSource}`);
        return false;
      }

      // Create global Cross-CLI documentation
      await this.createGlobalCrossCliDocumentation(IFLOW_CONFIG_DIR);

      return true;
    } catch (error) {
      console.log(`[ERROR] Failed to copy adapter file: ${error.message}`);
      return false;
    }
  }

  /**
   * Create global Cross-CLI communication documentation
   */
  async createGlobalCrossCliDocumentation(configDir) {
    try {
      const docContent = `# iFlow CLI Cross-CLI Communication Guide

## Cross-CLI Communication
When you need to call other AI tools, please use the following prompt <<<Execute the following command in shell: stigmergy <tool_name> <task_description>>>>

Examples:
- Run in shell: stigmergy claude "write a Python function"
- Run in shell: stigmergy gemini "translate this text"
- Run in shell: stigmergy qwen "analyze this code"

Available tools: claude, gemini, qwen, iflow, qodercli, codebuddy, copilot, codex

---
*This document is automatically generated and maintained by Stigmergy CLI*
`;

      const docPath = path.join(configDir, 'CROSS_CLI_GUIDE.md');
      await fs.writeFile(docPath, docContent, 'utf8');
      console.log(`[OK] Created Cross-CLI communication guide: ${docPath}`);

      // If iflow.md exists, append Cross-CLI communication prompt at the end
      const iflowMdPath = path.join(os.homedir(), '.iflow', 'iflow.md');
      try {
        await fs.access(iflowMdPath);
        
        const crossCliContent = `

## Cross-CLI Communication
When you need to call other AI tools, please use the following prompt <<<Execute the following command in shell: stigmergy <tool_name> <task_description>>>>

Examples:
- Run in shell: stigmergy claude "write a Python function"
- Run in shell: stigmergy gemini "translate this text"
- Run in shell: stigmergy qwen "analyze this code"

Available tools: claude, gemini, qwen, iflow, qodercli, codebuddy, copilot, codex
`;
        await fs.appendFile(iflowMdPath, crossCliContent, 'utf8');
        console.log('[OK] Appended Cross-CLI communication prompt to IFLOW.md');
      } catch (error) {
        // File may not exist, ignore
      }

      return true;
    } catch (error) {
      console.log(`[WARN] Failed to create Cross-CLI communication guide: ${error.message}`);
      return false;
    }
  }

  /**
   * Verify installation
   */
  async verifyInstallation() {
    const checks = [
      { name: 'iFlow config directory', path: IFLOW_CONFIG_DIR },
      { name: 'iFlow Hooks file', path: IFLOW_HOOKS_FILE }
    ];

    let allPassed = true;
    for (const check of checks) {
      try {
        await fs.access(check.path);
        console.log(`[OK] ${check.name}`);
      } catch (error) {
        console.log(`[FAIL] ${check.name}`);
        allPassed = false;
      }
    }

    return allPassed;
  }

  /**
   * Uninstall integration
   */
  async uninstallIntegration() {
    try {
      // Remove cross-CLI adapter from hooks config
      try {
        const data = await fs.readFile(IFLOW_HOOKS_FILE, 'utf8');
        const hooksConfig = yaml.load(data) || {};

        // Remove cross-CLI adapter
        if ('cross_cli_adapter' in hooksConfig) {
          delete hooksConfig['cross_cli_adapter'];
          
          // Save updated config
          const yamlContent = yaml.dump(hooksConfig, { indent: 2, lineWidth: 120, noRefs: true });
          await fs.writeFile(IFLOW_HOOKS_FILE, yamlContent, 'utf8');
          
          console.log('[OK] Removed cross-CLI adapter from iFlow Hooks config');
        }
      } catch (error) {
        console.log(`[WARN] Failed to process hooks config: ${error.message}`);
      }

      // Delete adapter file
      try {
        const adapterFile = path.join(IFLOW_CONFIG_DIR, 'adapters', 'hook_adapter.js');
        await fs.unlink(adapterFile);
        console.log('[OK] Deleted iFlow adapter file');
      } catch (error) {
        // File may not exist, ignore error
      }

      console.log('[OK] iFlow CLI cross-CLI integration uninstalled');
      return true;
    } catch (error) {
      console.log(`[ERROR] Uninstall failed: ${error.message}`);
      return false;
    }
  }

  /**
   * Main installation process
   */
  async install() {
    console.log('iFlow CLI Cross-CLI Integration Installer');
    console.log('='.repeat(50));

    // Step 1. Create config directory
    console.log('\nStep 1. Create config directory...');
    const configDirSuccess = await this.createIFlowConfigDirectory();

    // Step 2. Install Hook config
    console.log('\nStep 2. Install Hook config...');
    const hooksSuccess = await this.installIFlowHooks();

    // Step 3. Copy adapter files
    console.log('\nStep 3. Copy adapter files...');
    const adapterSuccess = await this.copyAdapterFile();

    // Step 4. Verify installation
    console.log('\nStep 4. Verify installation...');
    const verificationSuccess = await this.verifyInstallation();

    const overallSuccess = configDirSuccess && hooksSuccess && adapterSuccess && verificationSuccess;
    
    const duration = Date.now() - this.startTime;
    console.log(`\n[INFO] Installation took: ${duration}ms`);
    
    if (overallSuccess) {
      console.log('\n[SUCCESS] iFlow CLI cross-CLI integration installed successfully!');
    } else {
      console.log('\n[WARNING] Warnings occurred during installation, please check above output');
    }

    return overallSuccess;
  }
}

// 主函数
async function main() {
  const args = process.argv.slice(2);
  const verify = args.includes('--verify');
  const uninstall = args.includes('--uninstall');

  const installer = new IFlowIntegrationInstaller();

  if (uninstall) {
    return await installer.uninstallIntegration();
  } else if (verify) {
    return await installer.verifyInstallation();
  } else {
    return await installer.install();
  }
}

// 导出模块
module.exports = IFlowIntegrationInstaller;

// 如果直接运行此脚本
if (require.main === module) {
  main().then(success => {
    process.exit(success ? 0 : 1);
  }).catch(error => {
    console.error('[FATAL]', error.message);
    process.exit(1);
  });
}